doc/developer/design/20210707_qgm_sql_high_level_representation.md

```diff
@@ -0,0 +1,842 @@
+in SQL since no predicate can cross the non-preserving boundary).
+
+### Recursive CTEs support
+
+A recursive CTE is a CTE with a union with two branches where one of the branches references the CTE and the other
+one doesn't (representing the base case). This can be easily supported by the proposed implementation. Circular
+memory ownership issues are avoided by making the model own all the nodes.
+
+Any traversal of the query graph must keep a set/bitset of visited boxes since the same box can be ranged over by
+several quantifiers. The same set/bitset will prevent infinite loops/stack overflows when traversing a recursive
+query.
+
+### Testing and debuggability of the normalization transformations
+
+The normalization transformations will be mostly tested using `datadriven` unit tests that will take raw SQL
+queries as their input and a list of transformations to be applied, and will dump the content of the graph after
+applying each of them. The query graph will be dumped as a `dot` graph which, once rendered, looks like the
+ones in the appendix with examples.
+
+The `dot` graph generator will also be very handy for debugging transformations. Also, for better debuggability
+Jthe rules themselves should not drive the traversal of the graph, but just make explicit the type of traversal
+they require. The traversal of the graph will be driven by some external code, that could be instrumentalized for
+dumping the full graph before and after applying the rule to every node of the graph. That way, by just patching
+one method, we could follow step by step the evolution of a query graph as it goes through the different
+transformations.
+
+### Lowering
+
+The lowering code will have to be modified so that it takes a Query Graph as its input.
+
+When lowering a `Select` box we will first build the join within that box. To do so, we will first build a
+dependency graph where each quantifier is connected to the quantifiers it is correlated with. We will lower
+the quantifiers in dependency order, starting with the uncorrelated ones. When lowering
+a correlated quantifier, we will apply its plan to the result of the sub-join of all the quantifiers it is
+connected to, resulting in simpler dataflow plans for lateral joins.
+
+The dependency graph will be built by traversing the sub-graph of the input quantifiers collecting the
+column references from the current context.
+
+Note that the lowering code will no longer produce binary joins exclusively.
+
+After lowering the join, a `Map` operator will be added with the non-column expressions in the projection
+of the box. Then, a `Projection` will be added. If the box must enforce distinctness, the corresponding
+`Reduce` operator will be added. If the box has `LIMIT`/`OFFSET`/`ORDER BY` a `TopK` operator will
+then be added.
+
+### View inlining
+
+View inlining could be implemented as a normalization rule where only non-materialized views are inlined.
```

Comment:
Do you see view inlining being replaced by QGM immediately or more as potential future work? What would it imply for MIR?